internal/cmd/agent/controller.go

```diff
@@ -2,109 +2,246 @@ package agent
+	upstreamConfig, fleetNamespace, clusterName, err := loadRegistration(ctx, systemNamespace, localConfig)
 	if err != nil {
+		setupLog.Error(err, "unable to load registration and start manager")
 		return err
 	}
 
-	agentInfo, err := register.Get(ctx, namespace, kc)
+	// Start manager for upstream cluster, we do not use leader election
+	setupLog.Info("listening for changes on upstream cluster", "cluster", clusterName, "namespace", fleetNamespace)
+
+	metricsAddr := ":8080"
+	probeAddr := ":8081"
+
+	mgr, err := ctrl.NewManager(upstreamConfig, ctrl.Options{
+		Scheme:                 scheme,
+		Metrics:                metricsserver.Options{BindAddress: metricsAddr},
+		HealthProbeBindAddress: probeAddr,
+		LeaderElection:         false,
+		// only watch resources in the fleet namespace
+		Cache: cache.Options{
+			DefaultNamespaces: map[string]cache.Config{fleetNamespace: {}},
+		},
+	})
 	if err != nil {
-		setupLog.Error(err, "failed to get registration for upstream cluster")
+		setupLog.Error(err, "unable to start manager")
 		return err
 	}
 
-	fleetNamespace, _, err := agentInfo.ClientConfig.Namespace()
-	if err != nil {
-		setupLog.Error(err, "failed to get namespace from upstream cluster")
-		return err
-	}
+	localCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	fleetRESTConfig, err := agentInfo.ClientConfig.ClientConfig()
+	reconciler, err := newReconciler(ctx, localCtx, mgr, localConfig, systemNamespace, fleetNamespace, agentScope)
 	if err != nil {
-		setupLog.Error(err, "failed to get kubeconfig for upstream cluster")
+		setupLog.Error(err, "unable to setup bundledeployment reconciler")
 		return err
 	}
 
-	fleetMapper, mapper, discovery, err := newMappers(ctx, fleetRESTConfig, clientConfig)
-	if err != nil {
-		setupLog.Error(err, "failed to get mappers")
+	// Set up the bundledeployment reconciler
+	if err = (reconciler).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "BundleDeployment")
 		return err
 	}
+	//+kubebuilder:scaffold:builder
 
-	appCtx, err := controllers.NewAppContext(
-		fleetNamespace, namespace, agentInfo.ClusterNamespace, agentInfo.ClusterName,
-		fleetRESTConfig, clientConfig, fleetMapper, mapper, discovery)
-	if err != nil {
-		setupLog.Error(err, "failed to create app context")
+	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to set up health check")
 		return err
 	}
-
-	err = controllers.Register(ctx,
-		appCtx,
-		fleetNamespace, defaultNamespace,
-		agentScope)
-	if err != nil {
-		setupLog.Error(err, "failed to register controllers")
+	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to set up ready check")
 		return err
 	}
 
-	if err := appCtx.Start(ctx); err != nil {
-		setupLog.Error(err, "failed to start app context")
+	setupLog.Info("starting manager")
+	//if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
```

Comment:
Is signals handling already configured somewhere else?

Reply:
Took me a while to find it again: https://github.com/rancher/fleet/blob/fleet-agent-controller-runtime/cmd/fleetagent/main.go#L14

Reply:
So this commented line can go away?